DataDog/dd-trace-py
# feat(flask): trace streamed responses

## Description

In an instrumented flask application ddtrace does not include spans generated by a streamed response in the root `flask.request` span. Currently the `flask.request` span terminates before the first byte is streamed to the client. 

To address this issue this proposal wraps flask applications in a wsgi middleware. The `flask.request` span produced by this middleware will encompass both receiving a request and sending the response to a client.

<img width="995" alt="Screen Shot 2022-06-28 at 4 46 08 PM" src="https://user-images.githubusercontent.com/62392438/176284702-f71353af-4cf5-4904-9bff-d125e2161ee9.png">

## Checklist
- [x] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [x] Ensure tests are passing for affected code.
- [x] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.




## Motivation

Accurately measure the duration of flask request. Instrument streamed/buffered responses. 

## Design 

Use the _BaseWSGIMiddleware introduced in this change: https://github.com/DataDog/dd-trace-py/pull/3907. By extending this base class we will wrap the flask request, the flask wsgi_app, and the response in a span. 

Note - This is not a breaking change. The top level flask span `flask.request` will have the same tags and attributes (resource name, service name, error fields, etc). This PR introduces adds more visibility, it does not remove an existing functionality

## Testing strategy

1. create a flask app with an endpoint that returns a generator.
2. run the flask app and send a request to the endpoint
3. flush all spans to the agent
4. ensure the following spans are generated:
  - flask.request
  - flask.application 
  - flask.response



## Reviewer Checklist
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Add to milestone.
